File: backend/app/api/v1/reports.py
from datetime import datetime

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.dependencies import get_db
from app.core.responses import success
from app.services.report import ReportService

router = APIRouter()


@router.get("/dashboard")
def dashboard(db: Session = Depends(get_db)):
    service = ReportService(db)
    return success(service.dashboard_stats())


@router.get("/budgets-by-status")
def budgets_by_status(status: str, db: Session = Depends(get_db)):
    service = ReportService(db)
    return success(service.budgets_by_status(status))


@router.get("/work-orders-by-status")
def work_orders_by_status(status: str, db: Session = Depends(get_db)):
    service = ReportService(db)
    return success(service.work_orders_by_status(status))


@router.get("/monthly-sales")
def monthly_sales(year: int = Query(default=None), db: Session = Depends(get_db)):
    year = year or datetime.now().year
    service = ReportService(db)
    return success(service.monthly_sales(year))


@router.get("/most-used-materials")
def most_used_materials(limit: int = 10, db: Session = Depends(get_db)):
    service = ReportService(db)
    return success(service.most_used_materials(limit))

File: frontend/src/api/resources/reports.ts
import http from "../http";
import { wrap } from "../wrap";
import type { DashboardStats, Budget, WorkOrder } from "../../types";

export const reportsApi = {
  dashboard: () => wrap<DashboardStats>(() => http.get("/reports/dashboard")),
  budgetsByStatus: (status: string) => wrap<Budget[]>(() => http.get("/reports/budgets-by-status", { params: { status } })),
  workOrdersByStatus: (status: string) => wrap<WorkOrder[]>(() => http.get("/reports/work-orders-by-status", { params: { status } })),
  monthlySales: (year?: number) =>
    wrap<{ month: string; total: number; total_usd: number }[]>(() => http.get("/reports/monthly-sales", { params: year ? { year } : {} })),
  mostUsedMaterials: (limit = 10) =>
    wrap<{ name: string; usage_count: number }[]>(() => http.get("/reports/most-used-materials", { params: { limit } })),
};
